File: app/src/main/java/com/example/myapplication/SamplingScanActivity.java
package com.example.myapplication;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.provider.DocumentsContract;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.camera.view.PreviewView;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import com.example.myapplication.camera.QrScanner;
import com.example.myapplication.data.AssetRepository;
import com.example.myapplication.feedback.ScanFeedback;
import com.example.myapplication.logic.AssetIdFormat;
import com.example.myapplication.logic.ScanClassifier;
import com.example.myapplication.logic.ScannedTag;
import com.example.myapplication.model.Asset;
import com.example.myapplication.ui.ScanResultCard;
import com.example.myapplication.ui.ScannerOverlayView;
import com.example.myapplication.util.PermissionGuide;
import com.example.myapplication.util.PhotoEvidence;

import java.util.List;

public class SamplingScanActivity extends AppCompatActivity {

    private static final int REQ_CAMERA = 100;
    private static final long TOAST_COOLDOWN_MS = 1500;

    // ── Intent extras ────────────────────────────────────
    public static final String EXTRA_TARGET_ID    = "target_id";
    public static final String EXTRA_TARGET_NAME  = "target_name";

    // ── Result extras（回傳給 SamplingActivity）─────────
    public static final String RESULT_RAW = "raw"; // 原始 QR Code 字串

    private PreviewView previewView;
    private ScannerOverlayView scannerOverlay;
    private ScanResultCard resultCard;
    private TextView tvTargetId, tvTargetName, tvTargetMeta, tvHint, icTorch;
    private View btnTorch, btnCancel, awaitingOverlay;

    private String targetId;
    private String targetName;
    private long lastWrongScanToast = 0;
    private boolean finishing = false;   // 命中後延遲返回期間，忽略後續掃描

    private QrScanner scanner; // 相機 + 解碼管線（deep module）
    private ScanFeedback feedback; // 聲音＋震動回饋
    private View permissionDenied; // 權限被拒引導畫面
    private boolean cameraStarted = false;

    // 拍照存證（QR 破損）
    private Uri pendingPhotoUri;
    private String pendingPhotoName;
    private final ActivityResultLauncher<Uri> takePhoto =
            registerForActivityResult(new ActivityResultContracts.TakePicture(), success -> {
                if (Boolean.TRUE.equals(success)) {
                    String n = pendingPhotoName != null ? pendingPhotoName : "照片";
                    Toast.makeText(this, "已存證：" + n, Toast.LENGTH_SHORT).show();
                } else {
                    // 取消或失敗（TakePicture 無法區分）：刪掉剛建立的空檔
                    if (pendingPhotoUri != null) {
                        try {
                            DocumentsContract.deleteDocument(getContentResolver(), pendingPhotoUri);
                        } catch (Exception ignored) { }
                    }
                    Toast.makeText(this, "照片未儲存", Toast.LENGTH_SHORT).show();
                }
                pendingPhotoUri = null;
                pendingPhotoName = null;
            });

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sampling_scan);

        previewView     = findViewById(R.id.preview_view);
        scannerOverlay  = findViewById(R.id.scanner_overlay);
        resultCard      = findViewById(R.id.scan_result_card);
        tvTargetId      = findViewById(R.id.tv_target_id);
        tvTargetName    = findViewById(R.id.tv_target_name);
        tvTargetMeta    = findViewById(R.id.tv_target_meta);
        tvHint          = findViewById(R.id.tv_hint);
        icTorch         = findViewById(R.id.ic_torch);
        btnTorch        = findViewById(R.id.btn_torch);
        btnCancel       = findViewById(R.id.btn_cancel);
        awaitingOverlay = findViewById(R.id.awaiting_overlay);

        // 系統列（瀏海／導覽列）內距：以「絕對值」計算避免多次 dispatch 疊加。底部操作列
        // 以 paddingBottom 讓底色填到螢幕底；浮動的提示與結果卡跟導覽列高度上移，
        // 避免三按鈕導覽時被操作列擠壓／重疊。
        ViewCompat.setOnApplyWindowInsetsListener(
                findViewById(R.id.scan_root),
                (view, insets) -> {
                    Insets bars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
                    float d = getResources().getDisplayMetrics().density;

                    setTopMargin(findViewById(R.id.top_bar), bars.top);

                    View bar = findViewById(R.id.bottom_bar);
                    bar.setPadding((int) (8 * d), (int) (12 * d), (int) (8 * d), (int) (12 * d) + bars.bottom);

                    setBottomMargin(findViewById(R.id.tv_hint), (int) (112 * d) + bars.bottom);
                    setBottomMargin(findViewById(R.id.scan_result_card), (int) (104 * d) + bars.bottom);

                    return WindowInsetsCompat.CONSUMED;
                }
        );

        // 從 Intent 取得目標
        targetId = getIntent().getStringExtra(EXTRA_TARGET_ID);
        targetName = getIntent().getStringExtra(EXTRA_TARGET_NAME);

        if (targetId == null || targetId.isEmpty()) {
            Toast.makeText(this, "缺少目標資產", Toast.LENGTH_SHORT).show();
            finish();
            return;
        }
        tvTargetId.setText(targetId);
        tvTargetName.setText(targetName != null ? targetName : "");

        // 目標的部門·地點（供掃描員核對），從記憶體清單取
        Asset target = findTarget();
        if (target != null) {
            tvTargetMeta.setText(safe(target.department) + " · " + safe(target.location));
        }

        scanner = new QrScanner();
        feedback = new ScanFeedback(this);

        permissionDenied = findViewById(R.id.permission_denied);
        permissionDenied.findViewById(R.id.btn_open_settings)
                .setOnClickListener(v -> PermissionGuide.openAppSettings(this));

        findViewById(R.id.btn_capture).setOnClickListener(v -> capturePhoto());

        btnCancel.setOnClickListener(v -> {
            if (finishing) return;   // 命中延遲返回期間，別讓取消覆蓋 RESULT_OK
            setResult(RESULT_CANCELED);
            finish();
        });

        btnTorch.setOnClickListener(v -> toggleTorch());

        // 相機權限
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA)
                == PackageManager.PERMISSION_GRANTED) {
            startCamera();
        } else {
            ActivityCompat.requestPermissions(this,
                    new String[]{ Manifest.permission.CAMERA }, REQ_CAMERA);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQ_CAMERA
                && grantResults.length > 0
                && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            startCamera();
        } else {
            // 不直接關頁：顯示引導畫面，提供「前往設定」
            permissionDenied.setVisibility(View.VISIBLE);
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        // 從系統設定授權返回：若已取得權限則收起引導、啟動相機
        if (!cameraStarted
                && ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA)
                == PackageManager.PERMISSION_GRANTED) {
            permissionDenied.setVisibility(View.GONE);
            startCamera();
        }
    }

    private void startCamera() {
        if (cameraStarted) return;
        cameraStarted = true;
        scanner.bind(this, previewView, this::handleScanResult);
    }

    // QR 破損：拍整張標籤存到 CSV 同資料夾（檔名帶目前尋找的目標編號）
    private void capturePhoto() {
        String name = PhotoEvidence.buildFileName(targetId);
        Uri uri = PhotoEvidence.createInCsvFolder(this, name);
        if (uri == null) {
            Toast.makeText(this, "請先載入 CSV（需資料夾權限）", Toast.LENGTH_LONG).show();
            return;
        }
        pendingPhotoUri = uri;
        pendingPhotoName = name;
        takePhoto.launch(uri);
    }

    // ── 掃到結果 ────────────────────────────────────────
    private void handleScanResult(String raw) {
        if (finishing) return;
        ScannedTag tag = ScannedTag.parse(raw);
        String scannedId = tag.id;

        if (scannedId.equals(targetId)) {
            Asset target = findTarget();
            boolean matched = target != null && ScanClassifier.matches(target, tag);

            if (matched) {
                // ✅ 相符：綠卡自動消散，短暫停留後回傳（由抽盤主畫面落檔）
                finishing = true;
                if (scannerOverlay != null) scannerOverlay.flashSuccess();
                feedback.success();
                resultCard.showSuccess("✅ 相符", targetId + "　" + safe(targetName));
                returnRawDelayed(raw, 1200L);
            } else {
                // ⚠️ 目標不符（部門／地點）：紅卡列差異對比，確認才回傳落檔
                finishing = true;
                if (scannerOverlay != null) scannerOverlay.flashError();
                feedback.unmatched();
                if (target != null) {
                    setAwaiting(true);
                    resultCard.showUnmatched("⚠️ 部門或地點不相符", tag, target,
                            () -> returnRaw(raw),                       // 確認寫入不相符 → 回傳落檔
                            () -> { setAwaiting(false); finishing = false; });  // 略過 → 恢復掃描
                } else {
                    // 找不到目標資產（理論上不會發生）：直接回傳，交由主畫面處理
                    returnRaw(raw);
                }
            }
        } else if (AssetIdFormat.isValid(scannedId)) {
            // 掃到別的（成格式的）資產：橘卡提示、不寫入、繼續掃；不成格式的雜訊靜默忽略
            long now = System.currentTimeMillis();
            if (now - lastWrongScanToast > TOAST_COOLDOWN_MS) {
                lastWrongScanToast = now;
                if (scannerOverlay != null) scannerOverlay.flashError();
                resultCard.showWarning("請掃描指定財產", "目前尋找：" + targetId);
            }
        }
    }

    /** 待確認狀態：暗遮罩＋取景框紅角停線＋提示改字（A4）。 */
    private void setAwaiting(boolean awaiting) {
        if (awaitingOverlay != null) {
            awaitingOverlay.setVisibility(awaiting ? View.VISIBLE : View.GONE);
        }
        if (scannerOverlay != null) scannerOverlay.setAwaitingConfirm(awaiting);
        tvHint.setText(awaiting ? "請先處理待確認項目" : "對準指定財產 QR Code");
        tvHint.setTextColor(ContextCompat.getColor(this,
                awaiting ? R.color.scan_flash_error : R.color.text_on_primary));
    }

    /** 從記憶體清單找出目前目標資產（取其部門／地點以判定相符）。 */
    private Asset findTarget() {
        List<Asset> list = AssetRepository.getInstance().getAssets();
        if (list != null) {
            for (Asset a : list) {
                if (a != null && targetId.equals(a.id)) return a;
            }
        }
        return null;
    }

    private void returnRaw(String raw) {
        Intent intent = new Intent();
        intent.putExtra(RESULT_RAW, raw);
        setResult(RESULT_OK, intent);
        finish();
    }

    private void returnRawDelayed(String raw, long delayMs) {
        Intent intent = new Intent();
        intent.putExtra(RESULT_RAW, raw);
        setResult(RESULT_OK, intent);
        previewView.postDelayed(this::finish, delayMs);
    }

    private static String safe(String s) {
        return s == null ? "" : s;
    }

    private static void setTopMargin(View v, int px) {
        ViewGroup.MarginLayoutParams lp = (ViewGroup.MarginLayoutParams) v.getLayoutParams();
        lp.topMargin = px;
        v.setLayoutParams(lp);
    }

    private static void setBottomMargin(View v, int px) {
        ViewGroup.MarginLayoutParams lp = (ViewGroup.MarginLayoutParams) v.getLayoutParams();
        lp.bottomMargin = px;
        v.setLayoutParams(lp);
    }

    // ── 手電筒 ───────────────────────────────────────────
    private void toggleTorch() {
        if (!scanner.hasFlashUnit()) {
            Toast.makeText(this, "此裝置不支援手電筒", Toast.LENGTH_SHORT).show();
            return;
        }
        boolean on = scanner.toggleTorch();
        icTorch.setText(on ? "💡" : "🔦");
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (scanner != null) scanner.close();
        if (feedback != null) feedback.release();
    }
}

File: app/src/main/java/com/example/myapplication/SamplingScanActivity.kt
package com.example.myapplication

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.Vibrator
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.camera.core.Camera
import androidx.camera.core.CameraSelector
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageProxy
import androidx.camera.core.Preview
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.camera.view.PreviewView
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.example.myapplication.logic.AssetIdFormat
import com.google.zxing.BinaryBitmap
import com.google.zxing.DecodeHintType
import com.google.zxing.MultiFormatReader
import com.google.zxing.PlanarYUVLuminanceSource
import com.google.zxing.common.HybridBinarizer
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * 抽盤掃描頁：只認「當前目標」那一張 QR，掃到別的成格式資產會提示（含冷卻）。
 *
 * Kotlin 筆記：
 * - Java 的 `public static final` 常數 → `companion object` 裡的 `const val`，Java 端仍可用
 *   `SamplingScanActivity.RESULT_RAW` 存取（給還沒轉 Kotlin 的呼叫端用）。
 * - `layoutParams as ViewGroup.MarginLayoutParams`：Kotlin 的型別轉換用 `as`。
 * - `getSystemService(Vibrator::class.java)` 用型別安全的多載，省掉 Java 的手動 cast。
 */
class SamplingScanActivity : AppCompatActivity() {

    private lateinit var previewView: PreviewView
    private lateinit var tvTargetId: TextView
    private lateinit var tvTargetName: TextView
    private lateinit var tvHint: TextView
    private lateinit var btnTorch: TextView
    private lateinit var btnCancel: Button

    private var targetId: String? = null
    private var lastWrongScanToast = 0L

    private var camera: Camera? = null
    private var torchOn = false
    private lateinit var cameraExecutor: ExecutorService
    private lateinit var zxingReader: MultiFormatReader

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sampling_scan)

        previewView = findViewById(R.id.preview_view)
        tvTargetId = findViewById(R.id.tv_target_id)
        tvTargetName = findViewById(R.id.tv_target_name)
        tvHint = findViewById(R.id.tv_hint)
        btnTorch = findViewById(R.id.btn_torch)
        btnCancel = findViewById(R.id.btn_cancel)

        // 處理瀏海：頂部資訊條和底部取消按鈕往內推
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.scan_root)) { _, insets ->
            val bars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            val density = resources.displayMetrics.density

            // 頂部資訊條加上瀏海高度
            val topBar = findViewById<View>(R.id.top_bar)
            (topBar.layoutParams as ViewGroup.MarginLayoutParams).let {
                it.topMargin = bars.top
                topBar.layoutParams = it
            }

            // 底部取消按鈕加上導覽鍵高度
            val cancel = findViewById<View>(R.id.btn_cancel)
            (cancel.layoutParams as ViewGroup.MarginLayoutParams).let {
                it.bottomMargin = bars.bottom + 24 * density.toInt()
                cancel.layoutParams = it
            }

            // 手電筒按鈕也跟著頂部偏移
            val torch = findViewById<View>(R.id.btn_torch)
            (torch.layoutParams as ViewGroup.MarginLayoutParams).let {
                it.topMargin = bars.top + 50 * density.toInt()
                torch.layoutParams = it
            }

            WindowInsetsCompat.CONSUMED
        }

        // 從 Intent 取得目標
        targetId = intent.getStringExtra(EXTRA_TARGET_ID)
        val targetName = intent.getStringExtra(EXTRA_TARGET_NAME)

        if (targetId.isNullOrEmpty()) {
            Toast.makeText(this, "缺少目標資產", Toast.LENGTH_SHORT).show()
            finish()
            return
        }
        tvTargetId.text = targetId
        tvTargetName.text = targetName ?: ""

        cameraExecutor = Executors.newSingleThreadExecutor()
        zxingReader = MultiFormatReader().apply {
            setHints(mapOf(DecodeHintType.TRY_HARDER to true))
        }

        btnCancel.setOnClickListener {
            setResult(Activity.RESULT_CANCELED)
            finish()
        }

        btnTorch.setOnClickListener { toggleTorch() }

        // 相機權限
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA)
            == PackageManager.PERMISSION_GRANTED
        ) {
            startCamera()
        } else {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.CAMERA), REQ_CAMERA)
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray,
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == REQ_CAMERA &&
            grantResults.isNotEmpty() &&
            grantResults[0] == PackageManager.PERMISSION_GRANTED
        ) {
            startCamera()
        } else {
            Toast.makeText(this, "需要相機權限才能掃描", Toast.LENGTH_LONG).show()
            finish()
        }
    }

    // ── 啟動相機 ─────────────────────────────────────────
    private fun startCamera() {
        val future = ProcessCameraProvider.getInstance(this)
        future.addListener({
            try {
                val provider = future.get()

                val preview = Preview.Builder().build()
                preview.setSurfaceProvider(previewView.surfaceProvider)

                val analysis = ImageAnalysis.Builder()
                    .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
                    .build()
                analysis.setAnalyzer(cameraExecutor, ::analyzeImage)

                provider.unbindAll()
                camera = provider.bindToLifecycle(
                    this, CameraSelector.DEFAULT_BACK_CAMERA, preview, analysis,
                )
            } catch (e: Exception) {
                Log.e(TAG, "啟動相機失敗", e)
            }
        }, ContextCompat.getMainExecutor(this))
    }

    // ── 解碼 ─────────────────────────────────────────────
    private fun analyzeImage(imageProxy: ImageProxy) {
        try {
            val buffer = imageProxy.planes[0].buffer
            val bytes = ByteArray(buffer.remaining())
            buffer.get(bytes)

            val width = imageProxy.width
            val height = imageProxy.height

            val source = PlanarYUVLuminanceSource(bytes, width, height, 0, 0, width, height, false)
            val bitmap = BinaryBitmap(HybridBinarizer(source))

            try {
                val result = zxingReader.decodeWithState(bitmap)
                val raw = result.text
                runOnUiThread { handleScanResult(raw) }
            } catch (decodeError: Exception) {
                // 沒掃到，正常情況
            } finally {
                zxingReader.reset()
            }
        } finally {
            imageProxy.close()
        }
    }

    // ── 掃到結果 ────────────────────────────────────────
    private fun handleScanResult(raw: String) {
        val scannedId = raw.split(";")[0].trim()

        if (scannedId == targetId) {
            // ✅ 正確的資產，回傳結果
            val intent = Intent()
            intent.putExtra(RESULT_RAW, raw)
            setResult(Activity.RESULT_OK, intent)
            navigatorVibrate()
            finish()
        } else if (AssetIdFormat.isValid(scannedId)) {
            // ❌ 是別的（成格式的）資產，才提示；不成格式的雜訊視為誤觸，靜默忽略
            val now = System.currentTimeMillis()
            if (now - lastWrongScanToast > TOAST_COOLDOWN_MS) {
                lastWrongScanToast = now
                Toast.makeText(this, "⚠️ 這不是當前要找的資產（$scannedId）", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun navigatorVibrate() {
        // 震動回饋
        val vibrator = getSystemService(Vibrator::class.java)
        vibrator?.vibrate(60)
    }

    // ── 手電筒 ───────────────────────────────────────────
    private fun toggleTorch() {
        val camera = camera ?: return
        if (!camera.cameraInfo.hasFlashUnit()) {
            Toast.makeText(this, "此裝置不支援手電筒", Toast.LENGTH_SHORT).show()
            return
        }
        torchOn = !torchOn
        camera.cameraControl.enableTorch(torchOn)
        btnTorch.text = if (torchOn) "💡" else "🔦"
    }

    override fun onDestroy() {
        super.onDestroy()
        if (::cameraExecutor.isInitialized) cameraExecutor.shutdown()
    }

    companion object {
        private const val TAG = "SamplingScanActivity"
        private const val REQ_CAMERA = 100
        private const val TOAST_COOLDOWN_MS = 1500L

        // ── Intent extras ────────────────────────────────
        const val EXTRA_TARGET_ID = "target_id"
        const val EXTRA_TARGET_NAME = "target_name"

        // ── Result extras（回傳給 SamplingActivity）─────────
        const val RESULT_RAW = "raw" // 原始 QR Code 字串
    }
}
